cmd/clusterctl/cmd/get_kubeconfig.go

```diff
@@ -40,11 +41,14 @@ var getKubeconfigCmd = &cobra.Command{
 		Gets the kubeconfig file for accessing a workload cluster`),
 
 	Example: Examples(`
-		# Get the workload cluster's kubeconfig.
+		# Get the workload cluster's user kubeconfig. If it is not available, then the workload cluster's system kubeconfig is returned.
```

Comment:
Is it intentional that we change the default behavior of get kubeconfig?

Reply:
I think the fallback is not implemented?

Reply:
It was previously discussed top make the default behavior to get the user kubeconfig. If the user config is not available then the command should download the workload cluster's system kubeconfig. Ref previous discussion here: https://github.com/kubernetes-sigs/cluster-api/issues/3661#issuecomment-694747492

And looke like the fallback is already implemented.

Reply:
Where is the fallback implemented? Not sure how I'm missing it but I went over the PR again, I can't find it.

Are we sure this automagic fallback is a good idea? Essentially this means you run a clusterctl command and can never be sure if you end up with a user or system kubeconfig.

Reply:
To be clear with fallback I understand: `clusterctl get kubeconfig` is run and it will return if it exists the user kubeconfig and if it doesn't exist the system kubeconfig.

Reply:
But in my opinion the current implementation as I understand it (no automatically returning a different kubeconfig if no user kubeconfig exists) is what we should do, the comment just doesn't reflect it.

Reply:
Found it ...

Do we really think this is a good idea? 

If I want to get a user kubeconfig and it doesn't exist I will just get the system kubeconfig automatically. That sounds pretty dangerous. I think we can't rely on some output signalling it given that `clusterctl get kubeconfig` is likely to be also run inside of some pipelines where nobody will notice a log line that clusterctl just returned another kubeconfig.

Reply:
Just discussed it with Yuvaraj, we would want to discuss this with Fabrizio first (he will be back next week)

Reply:
Hi @fabriziopandini, looping you into this thread since we're waiting on you to finalize the behavior on this command.

Reply:
> Found it ...
> 
> I'm not sure if it is a good idea to be honest
> 
> If I want to get a user kubeconfig and it doesn't exist (or doesn't exist yet) I will just get the system kubeconfig automatically. That sounds pretty dangerous (I'm assuming a user kubeconfig doesn't have the same full access as the system kubeconfig). I think we can't rely on some output signalling it given that `clusterctl get kubeconfig` is likely to be also run inside of some pipelines where nobody will notice a log line that clusterctl just returned another kubeconfig.

@fabriziopandini Do you think we should change the default behavior?

Reply:
@sbueringer - so are you thinking if you request the user kubeconfig and it doesn't exist (and perhaps never will for a specific provider) that it returns an error message / specific exit code?

Reply:
I think that Stefan expressed a valid concern, given that falling back from a kubeconfig to another can lead to security issues (if the two kubeconfig have higher RBAC premissions)

A possible way forward is that `clusterctl get kubeconfig` could support retrieving a different kubeconfig only if this is specifically requested, e.g. `clusterctl get kubeconfig --type=user`.

However, I have some concerns about this approach as well because:
- It requires providing EKS-specific instructions to the users, and with this regards it is not much different than the current situation where we are giving instruction to retrieve the user kubeconfig via clusterctl
- It could give users the impression that clusterctl is a tool to manage many kubeconfig, which is not something we should do IMO.

The only alternative is to define a strict contract for providers (and BTW, no matter the decision I do expect this PR to extend the contract for control plane providers) that requires that control plane providers should create kubeconfig in a well-defined order (first user-kubeconfig, the defualt kubeconfig).
TBH I do not know if this is acceptable as idea given that there is still a chance of privilege escalation if  for any reason the user kubeconfig gets deleted

Reply:
Discussed in Aug 3rd Office hours
We are open to idea on how to make UX nice but we cannot trade on security concern
Also, it will be great if we can discuss this in terms of contract all the control plane providers should abide to.